File: neet_code_all_cpp/array_hash/num_of_interchangeable_rectangle.cpp
#include <iostream>
#include <vector>
#include <map>

class Solution {
    public:
        // AC
        long long interchangeableRectangles(std::vector<std::vector<int>>& rectangles) {
            std::map<std::vector<int>, std::vector<int>> map;
            for(int i = 0; i < rectangles.size(); ++i) {
                auto gcd_rct = gcd(rectangles[i][0], rectangles[i][1]);
                auto rct_reduced = std::vector{rectangles[i][0] / gcd_rct, rectangles[i][1] / gcd_rct};

                if(map.find(rct_reduced) != map.end()) {
                    map[rct_reduced].push_back(i);
                } else {
                    map.insert({rct_reduced, std::vector{i}});
                }

            }

            long long result = 0;
            for(auto& [v, idx_list]: map) {
                auto n = (long long)idx_list.size();
                result += n * (n - 1) / 2;
            }

            return result;
        }

        // AC
        long long interchangeableRectangles2(std::vector<std::vector<int>>& rectangles) {

            std::map<std::pair<int, int>, long long> map;
            for(int i = 0; i < rectangles.size(); ++i) {
                auto gcd_rct = gcd(rectangles[i][0], rectangles[i][1]);
                auto rct_reduced = std::pair{rectangles[i][0] / gcd_rct, rectangles[i][1] / gcd_rct};

                ++map[rct_reduced];
            }

            long long result = 0;
            for(auto& [v, n]: map) {
                result += n * (n - 1) / 2;
            }

            return result;
        }

        int gcd(int a, int b) {
            if(b == 0) {
                return a;
            }

            return gcd(b, a % b);
        }
};

// 模範解答
class SolutionAns {
    public:
        long long interchangeableRectangles(std::vector<std::vector<int>>& rectangles) {
            std::map<long double, int> hash;
            long long result = 0;

            for(int i = 0; i < rectangles.size(); ++i) {
                auto ratio = (long double)(rectangles[i][0]) / (long double)(rectangles[i][1]);

                if(hash.find(ratio) != hash.end()) {
                    ++hash[ratio];
                } else {
                    hash[ratio] = 1;
                }

            }

            for(auto& [v, num]: hash) {
                result += (long long)(num) * (long long)(num - 1) / 2;
            }

            return result;
        }
};

int main(void) {
    using Case = std::vector<std::vector<int>>;

    Case case_1 = {{4,8},{3,6},{10,20},{15,30}}; 
    // => 6
    Case case_2 = {{4,5},{7,8}};
    // => 0

    Solution s_1;

    std::cout << s_1.interchangeableRectangles(case_1) << std::endl;
    std::cout << s_1.interchangeableRectangles(case_2) << std::endl;

    std::cout << s_1.interchangeableRectangles2(case_1) << std::endl;
    std::cout << s_1.interchangeableRectangles2(case_2) << std::endl;

    SolutionAns s_ans;

    std::cout << s_ans.interchangeableRectangles(case_1) << std::endl;
    std::cout << s_ans.interchangeableRectangles(case_2) << std::endl;
}

File: neet_code_all/src/array_hash/num_of_interchangeable_rectangle.rs
use std::collections::HashMap;
use std::collections::HashSet;

struct Solution {}
impl Solution {
    // TLE
    pub fn interchangeable_rectangles(rectangles: Vec<Vec<i32>>) -> i64 {
        let mut rct_list = rectangles;

        for rct in rct_list.iter_mut() {
            let gcd_rct = Self::gcd(rct[0], rct[1]);
            rct[0] /= gcd_rct;
            rct[1] /= gcd_rct;
        }
        let mut result = 0;

        for i in 0..rct_list.len() {
            let mut set = HashSet::new();
            set.insert(rct_list[i].clone());
            for j in i + 1..rct_list.len() {
                if set.contains(&rct_list[j]) {
                    result += 1;
                }
            }
            set.remove(&rct_list[i]);
        }

        result
    }

    // AC
    pub fn interchangeable_rectangles_2(rectangles: Vec<Vec<i32>>) -> i64 {
        let rct_list = rectangles;

        let mut map: HashMap<Vec<i32>, Vec<usize>> = HashMap::new();
        for (i, rct) in rct_list.iter().enumerate() {
            let gcd_rct = Self::gcd(rct[0], rct[1]);
            let rct_reduced = vec![rct[0] / gcd_rct, rct[1] / gcd_rct];

            if let Some(map_key) = map.get_mut(&rct_reduced) {
                map_key.push(i);
            } else {
                map.insert(rct_reduced, vec![i]);
            }
        }
        //println!("map:\n {:?}", map);

        let mut result = 0;
        for (_v, idx_list) in map {
            let n = idx_list.len() as i64;
            result += n * (n - 1) / 2;
        }

        result
    }

    // AC(2をもう少しシンプルにした)
    pub fn interchangeable_rectangles_3(rectangles: Vec<Vec<i32>>) -> i64 {
        let rct_list = rectangles;

        let mut map: HashMap<(i32, i32), usize> = HashMap::new();
        for rct in rct_list.iter() {
            let gcd_rct = Self::gcd(rct[0], rct[1]);
            let rct_reduced = (rct[0] / gcd_rct, rct[1] / gcd_rct);

            *map.entry(rct_reduced).or_default() += 1;
        }
        //println!("map:\n {:?}", map);

        let mut result = 0;
        for (_v, num) in map {
            result += num * (num - 1) / 2;
        }

        result as i64
    }

    fn gcd(a: i32, b: i32) -> i32 {
        match b {
            0 => a,
            _ => Self::gcd(b, a % b),
        }
    }
}

// 模範解答
struct SolutionAns {}
impl SolutionAns {
    // AC
    pub fn interchangeable_rectangles(rectangles: Vec<Vec<i32>>) -> i64 {
        rectangles
            .into_iter()
            .map(|v| {
                let gcd = SolutionAns::compute_gcd(v[0], v[1]);
                (v[0] / gcd, v[1] / gcd)
            })
            .fold(HashMap::new(), |mut dict, t| {
                dict.entry(t).and_modify(|cnt| *cnt += 1).or_insert(1);
                dict
            })
            .into_iter()
            .filter(|(_, v)| *v > 1)
            .map(|(_, v)| v)
            .fold(0, |mut cnt, v| {
                cnt += v * (v - 1) / 2;
                cnt
            })
    }

    fn compute_gcd(mut a: i32, mut b: i32) -> i32 {
        while a > 0 && b > 0 {
            if a > b {
                a %= b;
            } else {
                b %= a;
            }
        }

        if a == 0 {
            b
        } else {
            a
        }
    }
}

fn main() {
    let case_1 = vec![vec![4, 8], vec![3, 6], vec![10, 20], vec![15, 30]];
    // => 6
    let case_2 = vec![vec![4, 5], vec![7, 8]];
    // => 0

    println!(
        "case_1: {}",
        Solution::interchangeable_rectangles(case_1.clone())
    );
    println!(
        "case_2: {}",
        Solution::interchangeable_rectangles(case_2.clone())
    );

    println!(
        "case_1: {}",
        Solution::interchangeable_rectangles_2(case_1.clone())
    );
    println!(
        "case_2: {}",
        Solution::interchangeable_rectangles_2(case_2.clone())
    );

    println!(
        "case_1: {}",
        Solution::interchangeable_rectangles_3(case_1.clone())
    );
    println!(
        "case_2: {}",
        Solution::interchangeable_rectangles_3(case_2.clone())
    );

    println!(
        "case_1: {}",
        SolutionAns::interchangeable_rectangles(case_1.clone())
    );
    println!(
        "case_2: {}",
        SolutionAns::interchangeable_rectangles(case_2.clone())
    );
}
